File: ProgrammingChallenges#Book/Chapter 1/CheckTheCheck/solutions/CheckTheCheck.py
from sys import stdin

class King:
    def __init__(self, x, y, isWhite):
        self.x = x
        self.y = y
        self.isWhite = isWhite


class Chess:
    def __init__(self, field, kingWhite, kingBlack):
        self.field = field
        self.kingWhite = kingWhite;
        self.kingBlack = kingBlack;
    

    def checkTheCheck(self):
        if (self.checkKing(self.kingWhite)):
            return "white king is in check."
        elif (self.checkKing(self.kingBlack)):
            return "black king is in check."
        return "no king is in check."
    
    
    def checkKing(self, king):
        return self.diagonalCheck(king) or self.straightCheck(king) or self.knightCheck(king)

    
    def diagonalCheck(self, king):
        # Verify top left
        for i, j in zip(range(king.y-1, -1, -1), range(king.x-1, -1, -1)):
            result = self.diagonalCheckValidation(king, i, j)
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify top right
        for i, j in zip(range(king.y-1, -1, -1), range(king.x+1, 8)):
            result = self.diagonalCheckValidation(king, i, j);
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify bottom left
        for i, j in zip(range(king.y+1, 8), range(king.x-1, -1, -1)):
            result = self.diagonalCheckValidation(king, i, j);
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify bottom right
        for i, j in zip(range(king.y+1, 8), range(king.x+1, 8)):
            result = self.diagonalCheckValidation(king, i, j);
            if (result == -1):
                break
            elif (result == 1):
                return True

        return False;
    
    def diagonalCheckValidation(self, king, i, j):
        if (self.field[i][j] == '.'):
            return 0
        else:
            # Verify if another piece with same color is on the way. 97 is ascii number for 'a' simbol
            if ((king.isWhite and ord(self.field[i][j]) < 97) or ((not king.isWhite) and ord(self.field[i][j]) >= 97)):
                return -1
            # Check pawn
            if ((king.isWhite and self.field[i][j] == 'p' and king.y - i == 1) or ((not king.isWhite) and self.field[i][j] == 'P' and i - king.y == 1)):
                return 1
            
            # Check bishop and queen
            if (self.field[i][j] == 'b' or self.field[i][j] == 'q' or self.field[i][j] == 'B' or self.field[i][j] == 'Q'):
                return 1
        
        return -1
    
    def straightCheck(self, king):
        # Verify top
        j = king.x
        for i in range(king.y-1, -1, -1):
            result = self.straightCheckValidation(king, i, j)
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify bottom
        j = king.x;
        for i in range(king.y+1, 8):
            result = self.straightCheckValidation(king, i, j)
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify left
        i = king.y
        for j in range(king.x-1, -1, -1):        
            result = self.straightCheckValidation(king, i, j)
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        # Verify right
        i = king.y
        for j in range(king.x+1, 8):        
            result = self.straightCheckValidation(king, i, j)
            if (result == -1):
                break
            elif (result == 1):
                return True
        
        return False;
    
    def straightCheckValidation(self, king, i, j):
        if (self.field[i][j] == '.'):
            return 0
        else:
            # Verify if another piece with same color is on the way. 97 is ascii number for 'a' simbol
            if ((king.isWhite and ord(self.field[i][j]) < 97) or ((not king.isWhite) and ord(self.field[i][j]) >= 97)):
                return -1

            # Check rook and queen
            if (self.field[i][j] == 'r' or self.field[i][j] == 'q' or self.field[i][j] == 'R' or self.field[i][j] == 'Q'):
                return 1
        
        return -1
    
    def knightCheck(self, king):
        # Verify top left
        i = king.y-1;
        j = king.x-2;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        i = king.y-2;
        j = king.x-1;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        # Verify top right
        i = king.y-2;
        j = king.x+1;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        i = king.y-1;
        j = king.x+2;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        # Verify bottom left
        i = king.y+1;
        j = king.x-2;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        i = king.y+2;
        j = king.x-1;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        # Verify bottom right
        i = king.y+2;
        j = king.x+1;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        i = king.y+1;
        j = king.x+2;
        if (self.knightCheckValidation(king, i, j)):
            return True
        
        return False
    
    def knightCheckValidation(self, king, i, j):
        if (i >= 0 and i < 8 and j >= 0 and j < 8):
            if ((king.isWhite and self.field[i][j] == 'n') or ((not king.isWhite) and self.field[i][j] == 'N')):
                return True
        
        return False


# Main program
def main():
    gameNumber = 1
    field = [['' for x in range(8)] for y in range(8)]
    while True:
        # Fill field
        isEmpty = 1
        for i in range(0, 8):
            inputLine = stdin.readline()
            for j in range (0, 8):
                field[i][j] = inputLine[j]
                if (field[i][j] != '.'):
                    isEmpty = 0
                if (field[i][j] == 'K'):
                    kingWhite = King(j, i, True)
                if (field[i][j] == 'k'):
                    kingBlack = King(j, i, False)

        if (isEmpty):
            break
        
        stdin.readline()

        # Verify chess
        chess = Chess(field, kingWhite, kingBlack)
        answer = chess.checkTheCheck()
        print("Game #%d: %s" % (gameNumber, answer))
        gameNumber = gameNumber + 1
        

main()

File: ProgrammingChallenges#Book/Chapter 1/CheckTheCheck/solutions/CheckTheCheck.c
#include <stdio.h>
#include <string.h>

struct King {
    int x;
    int y;
    int isWhite;
};

struct ChessField {
    struct King kingWhite;
    struct King kingBlack;
    char field[8][8];
};

struct ChessField chess;

void checkTheCheck(char *answer);
int checkKing(struct King king);
int diagonalCheck(struct King king);
int diagonalCheckValidation(struct King king, int i, int j);
int straightCheck(struct King king);
int straightCheckValidation(struct King king, int i, int j);
int knightCheck(struct King king);
int knightCheckValidation(struct King king, int i, int j);



/* Main program */
int main () {
    int i, j, isEmpty, gameNumber = 1;
    struct King kingWhite, kingBlack;
    char answer[100];
    char inputLine[10];

    while (1) {
        /* Fill field */            
        isEmpty = 1;
        for (i = 0; i < 8; i++) {
            fgets(inputLine, sizeof inputLine, stdin);
            for (j = 0; j < 8; j++) {
                chess.field[i][j] = inputLine[j];
                if (chess.field[i][j] != '.') {
                    isEmpty = 0;
                }
                if (chess.field[i][j] == 'K') {
                    chess.kingWhite.x = j;
                    chess.kingWhite.y = i;
                    chess.kingWhite.isWhite = 1;
                }
                if (chess.field[i][j] == 'k') {
                    chess.kingBlack.x = j;
                    chess.kingBlack.y = i;
                    chess.kingBlack.isWhite = 0;
                }
            }
        }

        if (isEmpty) {
            break;
        }
        fgets(inputLine, sizeof inputLine, stdin);

        /* Verify chess */
        checkTheCheck(answer);
        printf("Game #%d: %s\n", gameNumber, answer);
        gameNumber++;
    }

    return 0;
}


void checkTheCheck(char *answer) {
    if (checkKing(chess.kingWhite)) {
        strcpy(answer, "white king is in check.");
    } else if (checkKing(chess.kingBlack)) {
        strcpy(answer, "black king is in check.");
    } else {
        strcpy(answer, "no king is in check.");
    }
}

int checkKing(struct King king) {
    return diagonalCheck(king) || straightCheck(king) || knightCheck(king);
}

int diagonalCheck(struct King king) {
    int i, j;
    /* Verify top left */
    for (i = king.y-1, j = king.x-1; i >= 0 && j >= 0; i--, j--) {
        int result = diagonalCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify top right */
    for (i = king.y-1, j = king.x+1; i >= 0 && j < 8; i--, j++) {
        int result = diagonalCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify bottom left */
    for (i = king.y+1, j = king.x-1; i < 8 && j >= 0; i++, j--) {
        int result = diagonalCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify bottom right */
    for (i = king.y+1, j = king.x+1; i < 8 && j < 8; i++, j++) {
        int result = diagonalCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    return 0;
}

int diagonalCheckValidation(struct King king, int i, int j) {
    if (chess.field[i][j] == '.') {
        return 0;
    } else {
        /* Verify if another piece with same color is on the way. 97 is ascii number for 'a' simbol */
        if ((king.isWhite && chess.field[i][j] < 97) || (!king.isWhite && chess.field[i][j] >= 97)) {
            return -1;
        }
        /* Check pawn */
        if ((king.isWhite && chess.field[i][j] == 'p' && king.y - i == 1) || (!king.isWhite && chess.field[i][j] == 'P' && i - king.y == 1)) {
            return 1;
        }
        /* Check bishop && queen */
        if (chess.field[i][j] == 'b' || chess.field[i][j] == 'q' || chess.field[i][j] == 'B' || chess.field[i][j] == 'Q') {
            return 1;
        }
    }
    
    return -1;
}

int straightCheck(struct King king) {
    int i, j;
    /* Verify top */
    for (i = king.y-1, j = king.x; i >= 0; i--) {
        int result = straightCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify bottom */
    for (i = king.y+1, j = king.x; i < 8; i++) {
        int result = straightCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify left */
    for (i = king.y, j = king.x-1; j >= 0; j--) {
        int result = straightCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    /* Verify right */
    for (i = king.y, j = king.x+1; j < 8; j++) {
        int result = straightCheckValidation(king, i, j);
        if (result == -1) {
            break;
        } else if (result == 1) {
            return 1;
        }
    }
    
    return 0;
}

int straightCheckValidation(struct King king, int i, int j) {
    if (chess.field[i][j] == '.') {
        return 0;
    } else {
        /* Verify if another piece with same color is on the way. 97 is ascii number for 'a' simbol */
        if ((king.isWhite && chess.field[i][j] < 97) || (!king.isWhite && chess.field[i][j] >= 97)) {
            return -1;
        }
        /* Check rook && queen */
        if (chess.field[i][j] == 'r' || chess.field[i][j] == 'q' || chess.field[i][j] == 'R' || chess.field[i][j] == 'Q') {
            return 1;
        }
    }
    
    return -1;
}

int knightCheck(struct King king) {
    int i, j;
    /* Verify top left */
    i = king.y-1;
    j = king.x-2;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    i = king.y-2;
    j = king.x-1;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    /* Verify top right */
    i = king.y-2;
    j = king.x+1;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    i = king.y-1;
    j = king.x+2;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    /* Verify bottom left */
    i = king.y+1;
    j = king.x-2;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    i = king.y+2;
    j = king.x-1;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    
    /* Verify bottom right */
    i = king.y+2;
    j = king.x+1;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    i = king.y+1;
    j = king.x+2;
    if (knightCheckValidation(king, i, j)) {
        return 1;
    }
    
    return 0;
}

int knightCheckValidation(struct King king, int i, int j) {
    if (i >= 0 && i < 8 && j >= 0 && j < 8) {
        if ((king.isWhite && chess.field[i][j] == 'n') || (!king.isWhite && chess.field[i][j] == 'N')) {
            return 1;
        }
    }
    return 0;
}